akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala

```diff
@@ -133,6 +135,167 @@ import scala.util.control.NonFatal
+          matPromise.failure(new NeverMaterializedException(ex))
+          super.onUpstreamFailure(ex)
+        }
+      }
+
+      /**
+       * Called when the output port has received a pull, and therefore ready to emit an element, i.e. [[GraphStageLogic.push]]
+       * is now allowed to be called on this port.
+       */
+      override def onPull(): Unit = {
+        if(null != subSink){
+          //delegate to subSink
+          subSink.pull()
+        }
+        else if(accumulated.size < n) {
+          pull(in)
+        }
+        else if(accumulated.size == n) { //corner case for n = 0, can be handled in FlowOps
+          materializeFlow()
+        }
+      }
+
+      override def onDownstreamFinish(cause: Throwable): Unit = {
+        if(null != subSink){
+          subSink.cancel(cause)
+        } else {
+          materializeFlow()
+        }
+      }
+
+      def materializeFlow(): Unit = {
+        val prefix = accumulated.reverse
+        accumulated = Nil
+        val flow = try f(prefix) catch {
+          case NonFatal(ex) =>
+            val neverMaterializedEx = new NeverMaterializedException(ex)
+            matPromise.failure(neverMaterializedEx)
+            throw neverMaterializedEx
+        }
+        subSource = new SubSourceOutlet[In]("subSource")
+        subSource.setHandler{
+          new OutHandler {
+            /**
+             * Called when the output port has received a pull, and therefore ready to emit an element, i.e. [[GraphStageLogic.push]]
+             * is now allowed to be called on this port.
+             */
+            override def onPull(): Unit = {
+              if(!isClosed(in) && !hasBeenPulled(in)) {
+                pull(in)
+              }
+            }
+
+            override def onDownstreamFinish(cause: Throwable): Unit = {
+              if(!isClosed(in)) {
+                cancel(in, cause)
+              }
+              super.onDownstreamFinish(cause)
+            }
+          }
+        }
+        subSink = new SubSinkInlet[Out]("subSink")
+        subSink.setHandler{
+          new InHandler {
+            /**
+             * Called when the input port has a new element available. The actual element can be retrieved via the
+             * [[GraphStageLogic.grab]] method.
+             */
+            override def onPush(): Unit = {
+              push(out, subSink.grab())
+            }
+
+            override def onUpstreamFinish(): Unit = {
+              completeStage()
+              super.onUpstreamFinish()
+            }
+
+            override def onUpstreamFailure(ex: Throwable): Unit = {
+              failStage(ex)
+              super.onUpstreamFailure(ex)
+            }
+          }
+        }
+        val runnableGraph = Source.fromGraph(subSource.source)
+          .viaMat(flow)(Keep.right)
+          .to(subSink.sink)
+        val m = Try(interpreter.subFusingMaterializer.materialize(runnableGraph))
+        matPromise.complete(m)
+        m.get
```

Comment:
I wonder if it is really important to be able to discern failure to materialize from failure to create the flow with the types, if we skip that this could be cleaner with a single `try  { create and materialize and set up subflows } catch { case ex => fail-promise-and-stage }`

Reply:
hmm, makes sense.
care must be taken to prevent 'dangling' sub-source or sub-sink situations